src/unix/lwt_unix.ml

```diff
@@ -659,6 +659,114 @@ let write_string ch buf pos len =
   let buf = Bytes.unsafe_of_string buf in
   write ch buf pos len
 
+module IO_vectors =
+struct
+  type _bigarray =
+    (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.Array1.t
+
+  type _buffer =
+    | Bytes of Bytes.t
+    | Bigarray of _bigarray
+
+  type _io_vector =
+    {buffer : _buffer;
+     mutable offset : int;
+     mutable length : int}
+
+  (* This representation does not give constant amortized time append across all
+     possible operation sequences, but it does for expected typical usage, in
+     which some number of append operations is followed by some number of
+     flatten operations. *)
+  type t =
+    {mutable prefix : _io_vector list;
+     mutable reversed_suffix : _io_vector list;
+     mutable count : int}
+
+  let create () = {prefix = []; reversed_suffix = []; count = 0}
```

Comment:
I'm not sure it's worth the extra complexity in the interface, but an optional argument of `?bytes` and `?bigarrays` here could be used to initialise the iov in one call rather than multiple appends.  I could use that in [cstruct](https://github.com/mirage/ocaml-cstruct) to perform the write with less overhead from the API.